broker/endpoints/Patient.js

```diff
@@ -22,66 +28,112 @@ exports.read = (req, res) => {
-    let related_id = result.link.split("/")[1];
-    let resource = RelatedPerson.toFHIR({
-      patient: `Patient/${result.id}`,
-      relationship: patient.relationship,
-    });
-    resource.id = related_id;
-    await axios.put(`/RelatedPerson/${related_id}`, resource);
-  } else {
+  let createPromises = [];
+
+  if (!patient) {
     res.status(400).json({ error: "at least one patient should be provided" });
     return;
   }
 
-  // resources for members
-  for (patient of patients) {
-    let related = {
-      patient: `Patient/${response.Patient[0].id}`,
-      relationship: patient.relationship,
-    };
+  // Create head of household to be referenced by other members
+  let [head, promises] = await createPatient(patient);
 
-    promises.push(createPatientWithLink(patient, related));
+  for (patient of patients) {
+    createPromises.push(createPatient(patient, head));
   }
 
-  Promise.all(promises).then((results) => {
-    for (result of results) {
-      response.Patient.push(result);
-    }
-    res.json(response);
-  });
+  // Resolve promises and add new promises
+  Promise.all(createPromises)
+    .then((values) => {
+      promises = [].concat.apply(
+        promises,
+        values.map(([_, newPromises]) => newPromises)
+      );
+
+      // Resolve promises and return list of QR codes
+      Promise.all(promises).then((results) => {
+        let response = {
+          Patient: results.reduce((qrCodes, result) => {
+            if (result.qr_code) qrCodes.push(result.qr_code);
+            return qrCodes;
+          }, []),
+        };
+        res.json(response);
+      });
+    })
+    .catch((e) => res.status(400).json(e));
 }
 
-// create patient and link to relatedperson
-async function createPatientWithLink(patient, related) {
-  // create related person
+/**
+ * Create FHIR Patient resource and return a list
+ * containing the ID of the new resource and a list
+ * of promises that need to be resolved.
+ *
+ * @param {Patient information} patient
+ * @param {ID of head of household} head
+ */
+async function createPatient(patient, head) {
+  let related;
+  let promises = [];
+
+  // If head is not provided, create RelatedPerson resource first
+  // and update reference to patient later.
+  if (head == null) related = { resourceType: "RelatedPerson" };
+  else
+    related = {
+      patient: `Patient/${head}`,
+      relationship: patient.relationship,
+    };
+
+  // Create RelatedPerson resource
   let resource = RelatedPerson.toFHIR(related);
-  let related_id = (await axios.post(`/RelatedPerson`, resource)).data.id;
+  let relatedID = (await axios.post(`/RelatedPerson`, resource)).data.id;
 
-  // create patient pointing to related person
-  patient.link = [`RelatedPerson/${related_id}`];
+  // Create Patient resource with ID from PatientIdService
+  // and link to RelatedPerson
+  let id = (await PatientIdService.createPatientId()).patient_id;
+  promises.push(PatientIdService.createQrCodeForPatientId(id));
```

Comment:
In the case where Broker fails to create a patientID/QR Code, broker still creates the patient FHIR resource even when it hits this error.  Furthermore, it appears to replace the existing contents of the hapi database with the singular patient when it encounters this error.  I'm thinking we should update the logic somewhere around here to short circuit if that happens.

Reply:
Done